group-coordinator/src/main/java/org/apache/kafka/coordinator/group/PartitionMetadataClient.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group;
+
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Client interface for retrieving latest offsets for topic partitions.
+ */
+public interface PartitionMetadataClient extends AutoCloseable {
+    /**
+     * Lists the latest offsets for the provided topic partitions.
+     *
+     * @param topicPartitions A set of topic partitions.
+     * @return A map of topic partitions to the completableFuture of their latest offsets
+     */
+    Map<TopicPartition, CompletableFuture<Long>> listLatestOffsets(
```

Comment:
Shouldn't we have Map over `TopicIdPartition` i.e.
```suggestion
    Map<TopicIdPartition, CompletableFuture<Long>> listLatestOffsets(
```

Reply:
The topic ID is actually not required at all. So the `listLatestOffsets` method is responsible for 2 important things ->
1. Find the destination node for a particular topic partition, which is the leader broker for that partition. This information is retrieved using metadataCache, and the specific method for it requires only the topic name, not the id.
2. The ListOffsetsRequest is built and sent to the previously calculated destination node. The requestData object also requires only the name, not the ID. 

The PR for an implementation of PartitionMetadataClient is already created. https://github.com/apache/kafka/pull/20852

Maybe that can provide a better picture for this.

Reply:
In the future, I think we'll want to use topic ID, but the underlying RPC doesn't support topic ID yet. I'm happy for this to be based on topic name for now.